File: assertion/javascript/question4.js
_REPLACE_ME_WITH_DIRECTIVES_

_REPLACE_ME_WITH_SOLUTION_

function main() {
    const testCases = [
        {
            arguments: "findHeaterPower(100212373)",
            got: findHeaterPower(100212373),
            expected: 19
        }
    ];

    const workingAnswer = (n) => {
        const s = String(n);
        let sum = 0;
        for (let i = 0; i < s.length; i++) {
            sum += Number(s[i]);
        }
        return sum;
    };
    const randomNumber = (a, b) => Math.floor(Math.random() * (b - a + 1) + a);
    for (let i = 0; i < 9; i++) {
        const n = randomNumber(1000000000, 9999999999);
        const expected = workingAnswer(n);
        const got = findHeaterPower(n);
        const arguments = `findHeaterPower(${n})`;
        testCases.push({ got, expected, arguments });
    }

    for (let i = 0; i < testCases.length; i++) {
        const test = testCases[i];
        if (test.got == test.expected) {
            console.log(`# ${i + 1} PASSING`);
            console.log(`> ARGUMENTS ${test.arguments}`);
            console.log(`> EXPECTED ${test.expected}`);
            console.log(`> GOT ${test.got}`);
        } else {
            console.log(`# ${i + 1} FAILED`);
            console.log(`> ARGUMENTS ${test.arguments}`);
            console.log(`> EXPECTED ${test.expected}`);
            console.log(`> GOT ${test.got}`);
        }
    }
}

main();

File: assertion/cpp/question4.cpp
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <time.h>
#include <string>

_REPLACE_ME_WITH_DIRECTIVES_

int findHeaterPower(long power);

_REPLACE_ME_WITH_SOLUTION_

int __workingAnswer(long power)
{
    int result = 0;
    while (power != 0)
    {
        result += power % 10;
        power /= 10;
    }

    return result;
}

typedef struct TestCase
{
    int expected;
    int got;
    std::string arguments;
} TestCase;

// creates a random number between min and max
long __randomNumber(int min, long max)
{
    return (std::rand() % (max - min + 1)) + min;
}

int main()
{
    srand(time(0));

    std::vector<TestCase> testCases(1);
    testCases[0].expected = 19;
    testCases[0].got = findHeaterPower(100212373);
    testCases[0].arguments = "findHeaterPower(100212373)";

    for (int i = 0; i < 9; i++)
    {
        long n = __randomNumber(1000000000, 9999999999);
        TestCase testResult{};
        testResult.expected = __workingAnswer(n);
        testResult.got = findHeaterPower(n);
        testResult.arguments = "findHeaterPower(" + std::to_string(n) + ")";
        testCases.push_back(testResult);
    }

    for (unsigned int i = 0; i < testCases.size(); i++)
    {
        TestCase test = testCases.at(i);

        if (test.got == test.expected)
        {
            printf("# %d PASSING\n", i + 1);
        }
        else
        {
            printf("# %d FAILED\n", i + 1);
        }

        printf("> ARGUMENTS %s\n", test.arguments.c_str());
        printf("> EXPECTED %d\n", test.expected);
        printf("> GOT %d\n", test.got);
    }
    return 0;
}
